common_migration/find_common_migrations.py

```diff
@@ -0,0 +1,444 @@
+        (
+            node.value
+            for node in class_node.body
+            if isinstance(node, ast.Assign)
+            and len(node.targets) == 1
+            and isinstance(node.targets[0], ast.Name)
+            and cast(ast.Name, node.targets[0]).id == attribute_name
+            and isinstance(node.value, ast.List)
+        ),
+        None
+    )
+
+    if ast_list is None:
+        return []
+
+    migration_names = []
+
+    for element in ast_list.elts:
+        if not isinstance(element, ast.Tuple):
+            raise TypeError(f'Value is not tuple for `{attribute_name}`: {filename}')
+
+        possible_pair = tuple(
+            node.s
+            for node in element.elts
+            if isinstance(node, ast.Str)
+        )
+
+        if len(possible_pair) != 2:
+            raise ValueError(f'Invalid tuple for `{attribute_name}`: {filename}')
+
+        # Only consider values in the apps we care about.
+        if possible_pair[0] in app_name_list:
+            migration_names.append(cast(Tuple[str, str], possible_pair))
+
+    return migration_names
+
+
+def parse_migration_dependencies(
+    filename: str,
+    app_name_list: List[str],
+) -> Tuple[List[Tuple[str, str]], List[Tuple[str, str]]]:
+    """
+    Given a migration filename, return a pair of the migration name and
+    dependencies for the migration.
+    """
+    with open(filename) as migration_file:
+        parsed = ast.parse(migration_file.read())
+
+    class_node = next(
+        (
+            node
+            for node in parsed.body
+            if isinstance(node, ast.ClassDef)
+            and any(
+                base.attr == 'Migration'
+                for base in node.bases
+                if isinstance(base, ast.Attribute)
+            )
+        ),
+        None
+    )
+
+    if class_node is None:
+        raise SyntaxError(
+            f'Migration class is missing for migration: {filename}'
+        )
+
+    dependencies = _read_migration_tuples(
+        filename=filename,
+        app_name_list=app_name_list,
+        class_node=class_node,
+        attribute_name='dependencies',
+    )
+    replaces = _read_migration_tuples(
+        filename=filename,
+        app_name_list=app_name_list,
+        class_node=class_node,
+        attribute_name='replaces',
+    )
+
+    return dependencies, replaces
+
+
+def hash_node(node: MigrationNode, app_name: str) -> bytes:
+    """
+    Given any migration node, walk all the way up the tree for a given app and
+    produce a unique hash value that identifies the subgraph from the given
+    migration node.
+    """
+    app_migration_set = {
+        (current_node.app_name, current_node.name)
+        for current_node in
+        walk_up_nodes(node, app_name)
+    }
+
+    combined_string = ' '.join(
+        string
+        for key in sorted(app_migration_set)
+        for string in key
+    )
+
+    return (
+        hashlib.sha256(combined_string.encode('utf-8'))
+        .digest()
```

Comment:
I think `hexdigest()` would be more sensible to use here.

Reply:
I'd like to keep this as `bytes` so you can't mix it up with `str` values like the `name`.